proposals/p3407.md

```diff
@@ -0,0 +1,199 @@
+
+<!-- tocstop -->
+
+## Abstract
+
+-   Require namespace members be declared in the same name scope as the
+    namespace is declared.
+-   Allow binding patterns to directly declare names in namespaces.
+-   Disallow using different namespaces in the same binding pattern.
+
+## Problem
+
+While the trivial case of `var NS.a` seems to be supported as a consequence of
+proposal
+[#107: Code and name organization](https://github.com/carbon-language/carbon-lang/pull/107),
+there are a couple syntactic options when binding multiple names.
+
+Also, there's no clear decision around code such as:
+
+```
+namespace NS;
+class ClassT {
+  // Is this a class member accessed through `NS`, or a file scope member inside
+  // `NS`? What is its lifetime?
+  var NS.a = 0;
+}
+```
+
+This proposal mainly aims to remove these ambiguities.
+
+## Background
+
+Namespaces are covered in
+[code and name organization](/docs/design/code_and_name_organization/#namespaces).
+Binding patterns are covered in
+[pattern matching](/docs/design/pattern_matching.md#binding-patterns).
+
+There's some discussion of `var` in
+[values, variables, and pointers](/docs/design/values.md), but it's specific to
+locals. That doesn't address other use cases, such as globals or member
+variables.
+
+## Proposal
+
+When used to declare names in binding patterns, as in `var` or `let`, all names
+must be in the same namespace. Declaring members of a namespace inside of an
+unrelated scope is invalid.
+
+See the changes to
+[code and name organization](/docs/design/code_and_name_organization/#namespaces)
+for reference.
+
+## Rationale
+
+-   [Performance-critical software](/docs/project/goals.md#performance-critical-software)
+
+    -   Only allowing binding patterns to declare variables in the same
+        namespace allows for a guarantee that the memory can be sequentially
+        allocated. This allows for copy elision on initializers with function
+        calls.
+
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+
+    -   Requiring that declarations of multiple names use `NS.a` syntax is
+        consistent with the single variable case.
+    -   Requiring namespace members be declared while in the same name scope as
+        the namespace itself makes lifetimes clearer.
+
+## Alternatives considered
+
+### Allow prefixing a tuple binding pattern with a namespace
+
+We could use the namespace to prefix the binding tuple. For example:
+
+```carbon
+var NS.(a: i32, b: i32) = (3, 4);
+```
+
+It's rare that we would have a single statement declare multiple names. As a
+consequence, the separation of the namespace qualifier from the declared
+identifier might end up unique to this syntax. In that context, we prefer `NS.a`
+for consistency with other cases, such as `class NS.class`.
+
+### Allow mixing namespaces within a binding pattern
```

Comment:
This also seems to cover unrelated scopes?

```suggestion
### Allow using namespaces in unrelated scopes or mixing them within a binding pattern
```

Reply:
I think these are two separate issues...

Reply:
Sure, but then I'd separate out the rationale too. I think the static member thing is more attached to the unrelated scope nesting than mixing within a binding, although certainly combining the two exacerbates the rationale.

Reply:
I've split and rephrased the sections. Note I've also dropped the `static` mentions: I think my original intent with that wasn't working for you, and it didn't feel like the rephrasing needed it.

Reply:
I somewhat understood the `static` thing, but I think your new version is really nice and simple, so let's go with it. =D Thanks!!